frontend/dockerfile/dockerfile_test.go

```diff
@@ -65,9 +65,48 @@ func TestIntegration(t *testing.T) {
 		testPullScratch,
 		testSymlinkDestination,
 		testHTTPDockerfile,
+		testCopySymlinks,
 	})
 }
 
+func testCopySymlinks(t *testing.T, sb integration.Sandbox) {
+	t.Parallel()
+
+	dockerfile := []byte(`
+FROM scratch
+COPY foo /
+COPY sub/l* alllinks/
+`)
+
+	dir, err := tmpdir(
+		fstest.CreateFile("Dockerfile", dockerfile, 0600),
+		fstest.CreateFile("bar", []byte(`bar-contents`), 0600),
+		fstest.Symlink("bar", "foo"),
+		fstest.CreateDir("sub", 0700),
+		fstest.CreateFile("sub/lfile", []byte(`baz-contents`), 0600),
+		fstest.Symlink("subfile", "sub/l0"),
+		fstest.CreateFile("sub/subfile", []byte(`subfile-contents`), 0600),
+		fstest.Symlink("second", "sub/l1"),
+		fstest.Symlink("baz", "sub/second"),
+		fstest.CreateFile("sub/baz", []byte(`baz-contents`), 0600),
```

Comment:
Can we have a "zip slip" test: `fstest.Symlink("qux", "../../../../../../../../../../../etc/shadow")

Reply:
I think we need another fix for that. There is no actual breakout as copy happens in a container but I think `copy` doesn't currently evaluate the symlinks with the source directory as the root(instead it would happen in the context of the rootfs of temporary container). Or I could add a test for the `llb.Local` to check the broken symlink behavior independently of dockerfiles (although that should be covered by unit tests already).